src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs

```diff
@@ -67,61 +67,47 @@ protected override Task CleanupConnections()
-                {
-                    Log.DuplicateConnectionId(Logger, connectionId, null);
-                    throw new ArgumentException("ConnectionId already exists.");
-                }
+            var clientContext = new ClientContext(connectionId);
 
-                // Writing from the application to the service
-                _ = ProcessMessageAsync(connectionId);
+            if (_clientConnections.TryAdd(connectionId, clientContext))
+            {
+                _ = ProcessMessageAsync(clientContext);
+                return ForwardMessageToApplication(connectionId, openConnectionMessage);
             }
-            catch (Exception e)
+            else
             {
-                // Fail to write initial open connection message to channel
-                Log.ConnectedStartingFailed(Logger, connectionId, e);
-                // Close channel and notify client to close connection
-                clientContext.Output.TryComplete();
-                await WriteAsync(new CloseConnectionMessage(connectionId, e.Message));
+                Log.DuplicateConnectionId(Logger, connectionId, null);
+                return Task.CompletedTask;
             }
         }
 
-        protected override async Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
+        protected override Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
         {
-            var connectionId = closeConnectionMessage.ConnectionId;
-            if (_clientConnections.TryGetValue(connectionId, out var clientContext))
-            {
-                try
-                {
-                    await clientContext.Output.WriteAsync(closeConnectionMessage);
-                }
-                catch (Exception e)
-                {
-                    Log.FailToWriteMessageToApplication(Logger, connectionId, e);
-                }
-            }
+            return ForwardMessageToApplication(closeConnectionMessage.ConnectionId, closeConnectionMessage);
         }
 
-        protected override async Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        protected override Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        {
+            return ForwardMessageToApplication(connectionDataMessage.ConnectionId, connectionDataMessage);
+        }
+
+        private async Task ForwardMessageToApplication(string connectionId, ServiceMessage message)
         {
-            var connectionId = connectionDataMessage.ConnectionId;
             if (_clientConnections.TryGetValue(connectionId, out var clientContext))
             {
                 try
                 {
-                    await clientContext.Output.WriteAsync(connectionDataMessage);
+                    await clientContext.Output.WriteAsync(message);
                 }
                 catch (Exception e)
                 {
                     Log.FailToWriteMessageToApplication(Logger, connectionId, e);
```

Comment:
>FailToWriteMessageToApplication [](start = 24, length = 31)

Shall we add a message type in the log so it would be clear to understand which message fail to write?